<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head><title>R: Analyses of Protein Post-Translational Modifications</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>
<h1> Analyses of Protein Post-Translational Modifications
<img class="toplogo" src="../../../doc/html/Rlogo.svg" alt="[R logo]" />
</h1>
<hr/>
<div style="text-align: center;">
<a href="../../../doc/html/packages.html"><img class="arrow" src="../../../doc/html/left.jpg" alt="[Up]" /></a>
<a href="../../../doc/html/index.html"><img class="arrow" src="../../../doc/html/up.jpg" alt="[Top]" /></a>
</div><h2>Documentation for package &lsquo;ptm&rsquo; version 0.2.0</h2>

<ul><li><a href="../DESCRIPTION">DESCRIPTION file</a>.</li>
</ul>

<h2>Help Pages</h2>


<table width="100%">
<tr><td style="width: 25%;"><a href="dot-get.exepath.html">.get.exepath</a></td>
<td>Find Full Paths to Executables</td></tr>
<tr><td style="width: 25%;"><a href="dot-get.url.html">.get.url</a></td>
<td>Get Web Resource</td></tr>
<tr><td style="width: 25%;"><a href="aa.at.html">aa.at</a></td>
<td>Residue Found at the Requested Position</td></tr>
<tr><td style="width: 25%;"><a href="aa.comp.html">aa.comp</a></td>
<td>Amino Acid Composition</td></tr>
<tr><td style="width: 25%;"><a href="abundance.html">abundance</a></td>
<td>Protein Abundance Data</td></tr>
<tr><td style="width: 25%;"><a href="ac.scan.html">ac.scan</a></td>
<td>Scan a Protein in Search of Acetylation Sites</td></tr>
<tr><td style="width: 25%;"><a href="acc.dssp.html">acc.dssp</a></td>
<td>Compute Residue Accessibility and SASA</td></tr>
<tr><td style="width: 25%;"><a href="atom.dpx.html">atom.dpx</a></td>
<td>Atom Depth Analysis</td></tr>
<tr><td style="width: 25%;"><a href="background.go.html">background.go</a></td>
<td>Search GO Terms for Background Set</td></tr>
<tr><td style="width: 25%;"><a href="ball.html">ball</a></td>
<td>Search for Atoms Close to a Given Atom</td></tr>
<tr><td style="width: 25%;"><a href="compute.dssp.html">compute.dssp</a></td>
<td>Compute and Return a DSSP File</td></tr>
<tr><td style="width: 25%;"><a href="custom.aln.html">custom.aln</a></td>
<td>Download and Align Orthologous Sequences</td></tr>
<tr><td style="width: 25%;"><a href="ddG.profile.html">ddG.profile</a></td>
<td>Contribution of a given position to changes in stability</td></tr>
<tr><td style="width: 25%;"><a href="ddG.ptm.html">ddG.ptm</a></td>
<td>PDB Model and Change in Stability of a Modified Protein</td></tr>
<tr><td style="width: 25%;"><a href="dis.scan.html">dis.scan</a></td>
<td>Scan a Protein in Search of Disease-Related PTM Sites</td></tr>
<tr><td style="width: 25%;"><a href="dist2closest.html">dist2closest</a></td>
<td>Search and Compute the Distance to the Closest Aa</td></tr>
<tr><td style="width: 25%;"><a href="dpx.html">dpx</a></td>
<td>Atom Depth Analysis</td></tr>
<tr><td style="width: 25%;"><a href="env.extract.html">env.extract</a></td>
<td>Sequence Environment Around a Given Position</td></tr>
<tr><td style="width: 25%;"><a href="env.matrices.html">env.matrices</a></td>
<td>Environment Matrices</td></tr>
<tr><td style="width: 25%;"><a href="env.plot.html">env.plot</a></td>
<td>Differential Sequence Environment Plot</td></tr>
<tr><td style="width: 25%;"><a href="env.Ztest.html">env.Ztest</a></td>
<td>Preferred/Avoided Amino Acids Within an Environment</td></tr>
<tr><td style="width: 25%;"><a href="find.aaindex.html">find.aaindex</a></td>
<td>Find the Amino Acid Indexes</td></tr>
<tr><td style="width: 25%;"><a href="foldx.assembly.html">foldx.assembly</a></td>
<td>Compute Assembly Free Energy</td></tr>
<tr><td style="width: 25%;"><a href="foldx.mut.html">foldx.mut</a></td>
<td>Compute Changes in Stability (DDG)</td></tr>
<tr><td style="width: 25%;"><a href="foldx.stab.html">foldx.stab</a></td>
<td>Compute Folding Free Energy (DG)</td></tr>
<tr><td style="width: 25%;"><a href="get.area.html">get.area</a></td>
<td>Atomic Solvation Energies.</td></tr>
<tr><td style="width: 25%;"><a href="get.go.html">get.go</a></td>
<td>Get Gene Ontology Annotation</td></tr>
<tr><td style="width: 25%;"><a href="get.hssp.html">get.hssp</a></td>
<td>Get a HSSP File</td></tr>
<tr><td style="width: 25%;"><a href="get.seq.html">get.seq</a></td>
<td>Import a Protein Sequence from a Database</td></tr>
<tr><td style="width: 25%;"><a href="gl.scan.html">gl.scan</a></td>
<td>Scan a Protein in Search of OGlcNAc Sites</td></tr>
<tr><td style="width: 25%;"><a href="gorilla.html">gorilla</a></td>
<td>GO Enrichment Analysis</td></tr>
<tr><td style="width: 25%;"><a href="hdfisher.go.html">hdfisher.go</a></td>
<td>Hypothesis-driven Fisher Test</td></tr>
<tr><td style="width: 25%;"><a href="hmeto.html">hmeto</a></td>
<td>Human MetO sites oxidized by hydrogen peroxide treatment.</td></tr>
<tr><td style="width: 25%;"><a href="id.features.html">id.features</a></td>
<td>Features Related to the Protein Entry</td></tr>
<tr><td style="width: 25%;"><a href="id.mapping.html">id.mapping</a></td>
<td>Identifier Mapping</td></tr>
<tr><td style="width: 25%;"><a href="imutant.html">imutant</a></td>
<td>Compute Changes in Stability (DDG)</td></tr>
<tr><td style="width: 25%;"><a href="is.at.html">is.at</a></td>
<td>Check Residue a Fixed Position</td></tr>
<tr><td style="width: 25%;"><a href="list.hom.html">list.hom</a></td>
<td>Search Homologous Entries</td></tr>
<tr><td style="width: 25%;"><a href="me.scan.html">me.scan</a></td>
<td>Scan a Protein in Search of Methylation Sites</td></tr>
<tr><td style="width: 25%;"><a href="meto.list.html">meto.list</a></td>
<td>List Proteins Found in MetOSite Matching a Keyword</td></tr>
<tr><td style="width: 25%;"><a href="meto.scan.html">meto.scan</a></td>
<td>Scans a Protein in Search of MetO Sites</td></tr>
<tr><td style="width: 25%;"><a href="meto.search.html">meto.search</a></td>
<td>Search for Specific MetO Sites</td></tr>
<tr><td style="width: 25%;"><a href="mkdssp.html">mkdssp</a></td>
<td>Compute DSSP File Using an In-House Version of the DSSP Software</td></tr>
<tr><td style="width: 25%;"><a href="msa.html">msa</a></td>
<td>Multiple Sequence Alignment</td></tr>
<tr><td style="width: 25%;"><a href="net.go.html">net.go</a></td>
<td>Gene Ontology Network</td></tr>
<tr><td style="width: 25%;"><a href="ni.scan.html">ni.scan</a></td>
<td>Scan a Protein in Search of Nitration Sites</td></tr>
<tr><td style="width: 25%;"><a href="p.scan.html">p.scan</a></td>
<td>Scan a Protein in Search of Phosphosites</td></tr>
<tr><td style="width: 25%;"><a href="pairwise.dist.html">pairwise.dist</a></td>
<td>Compute Euclidean Distances</td></tr>
<tr><td style="width: 25%;"><a href="parse.dssp.html">parse.dssp</a></td>
<td>Parse a DSSP File to Return a Dataframe</td></tr>
<tr><td style="width: 25%;"><a href="parse.hssp.html">parse.hssp</a></td>
<td>Parse a HSSP File to Return Dataframes</td></tr>
<tr><td style="width: 25%;"><a href="pdb.chain.html">pdb.chain</a></td>
<td>Download and/or Split PDB Files.</td></tr>
<tr><td style="width: 25%;"><a href="pdb.pep.html">pdb.pep</a></td>
<td>Check Whether an Oligopeptide is Found in the PDB</td></tr>
<tr><td style="width: 25%;"><a href="pdb.quaternary.html">pdb.quaternary</a></td>
<td>Protein Subunit Composition</td></tr>
<tr><td style="width: 25%;"><a href="pdb.res.html">pdb.res</a></td>
<td>Check Whether a Given Residue is Found in the PDB</td></tr>
<tr><td style="width: 25%;"><a href="pdb.select.html">pdb.select</a></td>
<td>Select the PDB with the Optimal Coverage to the UniProt Sequence</td></tr>
<tr><td style="width: 25%;"><a href="pdb.seq.html">pdb.seq</a></td>
<td>Get Chain Sequences</td></tr>
<tr><td style="width: 25%;"><a href="pdb2uniprot.html">pdb2uniprot</a></td>
<td>Returns the UniProt ID Given the PDB and Chain IDs</td></tr>
<tr><td style="width: 25%;"><a href="prot2codon.html">prot2codon</a></td>
<td>Find the Coding Triplets for a Given Protein</td></tr>
<tr><td style="width: 25%;"><a href="ptm.plot.html">ptm.plot</a></td>
<td>Plot Values of a Property and PTM Sites Along the Protein Sequence</td></tr>
<tr><td style="width: 25%;"><a href="ptm.scan.html">ptm.scan</a></td>
<td>Scan a Protein in Search of PTM Sites</td></tr>
<tr><td style="width: 25%;"><a href="reg.scan.html">reg.scan</a></td>
<td>Scan a Protein in Search of Regulatory PTM Sites</td></tr>
<tr><td style="width: 25%;"><a href="renum.html">renum</a></td>
<td>Renumerate Residue Position</td></tr>
<tr><td style="width: 25%;"><a href="renum.meto.html">renum.meto</a></td>
<td>Renumerate Residue Position</td></tr>
<tr><td style="width: 25%;"><a href="renum.pdb.html">renum.pdb</a></td>
<td>Renumerate Residue Position</td></tr>
<tr><td style="width: 25%;"><a href="res.dist.html">res.dist</a></td>
<td>Compute Distances Between Residues</td></tr>
<tr><td style="width: 25%;"><a href="res.dpx.html">res.dpx</a></td>
<td>Residue Depth Analysis</td></tr>
<tr><td style="width: 25%;"><a href="saro.dist.html">saro.dist</a></td>
<td>Compute Distances to the Closest Aromatic Residues</td></tr>
<tr><td style="width: 25%;"><a href="saro.geometry.html">saro.geometry</a></td>
<td>Compute Geometric Parameters of S-Aromatic Motifs</td></tr>
<tr><td style="width: 25%;"><a href="saro.motif.html">saro.motif</a></td>
<td>Search for S-Aromatic Motifs</td></tr>
<tr><td style="width: 25%;"><a href="search.go.html">search.go</a></td>
<td>Search a Simple User Query</td></tr>
<tr><td style="width: 25%;"><a href="shannon.html">shannon</a></td>
<td>Compute Shannon Entropy</td></tr>
<tr><td style="width: 25%;"><a href="site.type.html">site.type</a></td>
<td>Compute Shannon Entropy and Sort out the Sites</td></tr>
<tr><td style="width: 25%;"><a href="sni.scan.html">sni.scan</a></td>
<td>Scan a Protein in Search of S-nitrosylation Sites</td></tr>
<tr><td style="width: 25%;"><a href="species.kegg.html">species.kegg</a></td>
<td>Convert Between Species Name and KEGG 3-Letter Code Format</td></tr>
<tr><td style="width: 25%;"><a href="species.mapping.html">species.mapping</a></td>
<td>Map Protein ID to Species</td></tr>
<tr><td style="width: 25%;"><a href="stru.part.html">stru.part</a></td>
<td>Partition of Structural Regions</td></tr>
<tr><td style="width: 25%;"><a href="su.scan.html">su.scan</a></td>
<td>Scan a Protein in Search of Sumoylation Sites</td></tr>
<tr><td style="width: 25%;"><a href="term.go.html">term.go</a></td>
<td>Get Core Information About the GO Term</td></tr>
<tr><td style="width: 25%;"><a href="ub.scan.html">ub.scan</a></td>
<td>Scan a Protein in Search of Ubiquitination Sites</td></tr>
<tr><td style="width: 25%;"><a href="uniprot2pdb.html">uniprot2pdb</a></td>
<td>Return the PDB and Chain IDs of the Provided UniProt Protein</td></tr>
<tr><td style="width: 25%;"><a href="xprod.html">xprod</a></td>
<td>Compute Cross Product</td></tr>
</table>
</body></html>
